raft topology: enable writes to previous CDC generations

patjed41 opened the pull request:
When we create a CDC generation and ring-delay is non-zero, the
timestamp of the new generation is in the future. Hence, we can
have multiple generations that can be written to. However, if we
add a new node to the cluster with the Raft-based topology, it
receives only the last committed generation. So, this node will
be rejecting writes considered correct by the other nodes until
the last committed generation starts operating.

In scylladb/scylladb#17134, we have allowed sending writes to the
previous CDC generations. So, the situation became even more
complicated. This PR adjusts the Raft-based topology
to ensure all required generations are loaded into memory and their
data isn't cleared too early.

To load all required generations into memory, we replace
`current_cdc_generation_{uuid, timestamp}` with the set containing
IDs of all committed generations - `committed_cdc_generations`.
To ensure this set doesn't grow endlessly, we remove an entry from
this set together with the data in CDC_GENERATIONS_V3.

Currently, we may clear a CDC generation's data from
CDC_GENERATIONS_V3 if it is not the last committed generation
and it is at least 24 hours old (according to the topology
coordinator's clock). However, after allowing writes to the
previous CDC generations, this condition became incorrect. We
might clear data of a generation that could still be written to.
The new solution introduced in this PR is to clear data of the
generations that finished operating more than 24 hours ago.

Apart from the changes mentioned above, this PR hardens
`test_cdc_generation_clearing.py`.

Fixes scylladb/scylladb#16916
Fixes scylladb/scylladb#17184
Fixes scylladb/scylladb#17288

patjed41 commented:
There is a trivial and harmless conflict in `service/topology_mutation.hh`. I'll ignore it for now because fixing it would also require rebasing #17134. I don't want to do it because it would slow down merging of that PR (CI would be restarted again).

After #17134 is merged, I will rebase anyway and fix it.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6740/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6740/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7961/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6740/)
#### Build Details:
- Duration: 1 hr 58 min
- Builder: spider4.cloudius-systems.com

scylladb-promoter commented:
## Docs Preview :book:
Docs Preview for this pull request is available [here](http://previews.docs.scylladb.com/scylladb/17374)

**Changed Files:**

**Note:** This preview will be available for 30 days and will be automatically deleted after that period. You can manually trigger a new build by committing changes.

kbr-scylla commented on db/system_keyspace.cc:
```diff
@@ -296,11 +296,6 @@ schema_ptr system_keyspace::cdc_generations_v3() {
              * range when the generation was first created. Together with the set of streams above it fully
```
> In `topology_coordinator:clean_obsolete_cdc_generations`, we compute the iterator of the last obsolete generation (`last_obsolete_gen_it`). For now, it is unnecessary, but we will use it in the following patch to clean `committed_cdc_generations` from obsolete generations.

But you do use it already in this patch, to remove data from CDC_GENERATIONS_V3.

patjed41 replied:
In this patch, I only need this: `auto id_upper_bound = (*last_obsolete_gen_it)->id;`. I don't need the `last_obsolete_gen_it` variable. I can find `id_upper_bound` directly in the loop.

Anyway, it looks like writing this paragraph wasn't a good idea since it's confusing and not important. I'll remove it.

kbr-scylla commented on test/topology_experimental_raft/test_cdc_generation_clearing.py:
```diff
@@ -63,13 +63,22 @@ async def tried_to_remove_new_gen() -> Optional[tuple[int, set[str], list[Host]]
-    logger.info(f"Generations after third clearing attempt: {gen_ids}")
-    assert len(gen_ids) == 1 and first_gen_id not in gen_ids and second_gen_id not in gen_ids
-    await check_system_topology_and_cdc_generations_v3_consistency(manager, hosts)
+    async with inject_error(manager.api, servers[0].ip_addr, "clean_obsolete_cdc_generations_change_ts_ub"):
```
Does this hardened test actually reproduce the bug https://github.com/scylladb/scylladb/issues/16916?

patjed41 replied:
Any test that expects two clearings would fail with this bug in the code. After the first clearing, CDC generation publisher got stuck due to an error. The old version of this test performed only one clearing and it was the last step of the test. So, the test passed even though the clearing didn't work properly.

Suppose this PR didn't exist and we didn't have to change the clearing logic. Then, I would send a direct fix for #16916 and harden the test in a similar way -- by expecting two clearings. I did exactly that on my local machine while investigating this issue. The hardened version passed only after the direct fix.

However, we have this PR and it changes the clearing logic. The bugged code is no longer here and the test is adjusted. Nevertheless, #16916 showed that the test should be stronger by expecting at least two clearings instead of one because the first one may do something wrong without being noticed. So, I've decided to harden the test anyway.

patjed41 commented:
v2:
- removed the paragraph about `last_obsolete_gen_it` from the fourth commit's message
- replaced `last_obsolete_gen_it` with `first_nonobsolete_gen_it`

patjed41 commented:
rebased

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6814/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6814/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8037/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6814/)
#### Build Details:
- Duration: 2 hr 2 min
- Builder: spider7.cloudius-systems.com

kbr-scylla commented on service/topology_coordinator.cc:
```diff
@@ -546,41 +546,48 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
-        mutation m = _sys_ks.make_cleanup_candidate_mutation(std::nullopt, mut_ts);
-
-        // Insert a tombstone covering all generations that have time UUID not higher than the candidate.
+        // Insert a tombstone covering all obsolete generations.
```
Hmm, I think it's theoretically possible that we may clean data of generation that was not published yet, which would be a big problem. At least I see no mechanism to protect us from this. We should probably stop our first_nonobsolete_gen_it search when we see that the gen is unpublished yet.

But IIUC this is a preexisting problem, it was present with the cleanup_candidate as well. And the current PR only improves the situation because it clears generations later. So we can fix in a follow-up, it should be a simple fix

patjed41 replied:
Yes, this is a preexisting problem. I'll send a follow-up.
